Add align measure pass

nkanazawa1989 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
This pass fixes the strange behavior of IBM Quantum backends when a circuit with length not multiple of 16 is provided. This behavior has been reported multiple times so far, especially some experiments requiring arbitrary delay between instructions, such as inversion recovery or Ramsey experiment. 

This malfunction is caused by the misalignment of acquire instruction from the clock boundary of 16dt, which causes a phase error in measurement channels, yielding the drift of blob position in the measurement IQ plane (you can check this behavior with level1 measurement).

Fix #6666

### Details and comments
Instead of fixing the rotation error, this PR adds a pass that shift the start time of measurement instruction to match with the 16dt clocks. This number will be provided by backend as `alignment`. The pass is currently called by the transpiler, but we can call this also from the IBMQ backend object via run method (in this case legacy backend can still cause the error). 

This needs https://github.com/Qiskit/qiskit-ibmq-provider/pull/983

```python
from qiskit import IBMQ, circuit, transpile

IBMQ.load_account()
backend = IBMQ.get_provider(hub='ibm-q', group='open', project='main').get_backend('ibmq_armonk')

qc = circuit.QuantumCircuit(1, 1)
qc.x(0)
qc.delay(110, 0, unit="dt")
qc.measure(0, 0)

qc.draw()  #

     ┌───┐┌────────────────┐┌─┐
q_0: ┤ X ├┤ Delay(110[dt]) ├┤M├
     └───┘└────────────────┘└╥┘
c: 1/════════════════════════╩═
                             0 

qct = transpile(qc, backend)

qct.draw()  #

     ┌───┐┌────────────────┐┌─┐
q_0: ┤ X ├┤ Delay(112[dt]) ├┤M├
     └───┘└────────────────┘└╥┘
c: 1/════════════════════════╩═
                             0 
```

# TODO
- [x] decide from where we call this pass
- [x] add reno

nkanazawa1989 commented:
This is the result of integration test

```python
from qiskit import IBMQ, circuit, transpile
import matplotlib.pyplot as plt
import numpy as np

IBMQ.load_account()
backend = IBMQ.get_provider(hub='ibm-q', group='open', project='main').get_backend("ibmq_armonk")

p_delay = circuit.Parameter("Δt")
delays = np.linspace(0, 100, 20)

qc = circuit.QuantumCircuit(1, 1)
qc.x(0)
qc.delay(p_delay, 0, unit="us")
qc.measure(0, 0)

qc.draw()  #

     ┌───┐┌───────────────┐┌─┐
q_0: ┤ X ├┤ Delay(Δt[us]) ├┤M├
     └───┘└───────────────┘└╥┘
c: 1/═══════════════════════╩═
                            0 

exp_qcs = [qc.assign_parameters({p_delay: delay}, inplace=False) for delay in delays]
result = backend.run(transpile(exp_qcs, backend), shots=256).result()

plt.plot(delays, [count.get("1", 0)/256 for count in result.get_counts()])
plt.xlabel("delay, us")
```
![image](https://user-images.githubusercontent.com/39517270/124146105-66539480-dac8-11eb-9a86-fadb2f0ef2bb.png)

ajavadia reviewed (commented):
partial review before i look at the actual pass algorithm

nkanazawa1989 commented on qiskit/transpiler/passes/scheduling/align_measures.py:
```diff
@@ -0,0 +1,127 @@
+
+        if len(dag.op_nodes(Delay)) == 0 or len(dag.op_nodes(Measure)) == 0:
+            # delay is only instruction that induce misalignment
+            # other instructions cannot be arbitrarily stretched
```
This instruction cannot be accepted by the backend because this already violate chunk size constraint. I was thinking to add chunk size validation for pulse gate later but I can add this check in this PR too.

nkanazawa1989 replied:
Just added custom gate validation example in this commit https://github.com/Qiskit/qiskit-terra/pull/6673/commits/72821a3881f8bef53262880a301872907c1ccc2e
I'll revert this commit and can move to separate PR if you wish.

nkanazawa1989 commented on qiskit/compiler/transpiler.py:
```diff
@@ -62,6 +62,7 @@ def transpile(
     pass_manager: Optional[PassManager] = None,
     callback: Optional[Callable[[BasePass, DAGCircuit, float, PropertySet, int], Any]] = None,
     output_name: Optional[Union[str, List[str]]] = None,
+    alignment: Optional[int] = None,
```
Thanks, resolved in https://github.com/Qiskit/qiskit-terra/pull/6673/commits/92163d40d11a83d4051e9876041609299bdd70ac

nkanazawa1989 commented on qiskit/compiler/transpiler.py:
```diff
@@ -186,6 +187,11 @@ def callback_func(**kwargs):
         output_name: A list with strings to identify the output circuits. The length of
             the list should be exactly the length of the ``circuits`` parameter.
+        alignment: An optional control hardware restriction on instruction time allocation.
+            The location of measure instruction is adjusted to be at quantized time that is
```
Thanks, resolved in https://github.com/Qiskit/qiskit-terra/pull/6673/commits/92163d40d11a83d4051e9876041609299bdd70ac

nkanazawa1989 commented on qiskit/compiler/transpiler.py:
```diff
@@ -355,6 +362,14 @@ def _transpile_circuit(circuit_config_tuple: Tuple[QuantumCircuit, Dict]) -> Qua
+            alignment=transpile_config["alignment"],
+            durations=pass_manager_config.instruction_durations,
+        )
+        pass_manager.append(_align_measure)
```
Now they are moved to generator functions. https://github.com/Qiskit/qiskit-terra/pull/6673/commits/92163d40d11a83d4051e9876041609299bdd70ac

nkanazawa1989 commented on test/python/transpiler/test_align_measure.py:
```diff
@@ -0,0 +1,130 @@
+        self.align_measure_pass = AlignMeasures(alignment=16, durations=instruction_durations)
+
+    def test_t1_experiment_type(self):
+        """Test T1 experiment type circuit."""
```
This is awesome idea! done in https://github.com/Qiskit/qiskit-terra/pull/6673/commits/92163d40d11a83d4051e9876041609299bdd70ac

nkanazawa1989 commented on qiskit/transpiler/passes/scheduling/align_measures.py:
```diff
@@ -0,0 +1,127 @@
+            # run ALAP schedule to find measurement times
+            for required_pass in scheduler.requires:
+                dag = required_pass.run(dag)
+            dag = scheduler.run(dag)
```
Thanks. I took the approach 1 with slight improvement of the logic. Because we don't want to reduce breaking change as much as possible, so new logic has validation of delay duration to trigger pass execution. If all delays are multiple of 16 this pass is not executed, thus calling transpile without scheduling option doesn't cause an error.

itoko replied:
I know, in general, we should not call another pass internally. I just want to note that we are accepting to break some existing codes using delays without scheduling, e.g. T1/T2 experiments, if requiring the external call of ALAP pass. They can be fixed only by supplying `scheduling_method` to `transpile` though. I'm wondering if this cost is allowable. What do you think? @ajavadia

ajavadia replied:
I think the existence of `alignment` should tell the transpiler to do scheduling, even if `scheduling_method` is not supplied. There is a similar issue here https://github.com/Qiskit/qiskit-terra/issues/6667

1ucian0 replied:
> but they only declare that a certain pass must have been run before them, not that a certain property must have been set. I think @1ucian0 has been meaning to implement this.

it is not possible to do "require a property" right now. However, you can conditionally run a pass based on the existence of a property which, if I understand this correctly, is the case in this situation.

nkanazawa1989 commented:
Thanks for the quick review. I updated the code according to your comments. The major changes are

- Update the logic of pass execution judgement. Just minimized the risk of unnecessarily executing the heavy pass. 
- Added example of pulse gate alignment execution https://github.com/Qiskit/qiskit-terra/pull/6673/commits/72821a3881f8bef53262880a301872907c1ccc2e. This can be moved to another PR, but in this way we can guarantee only delays can cause misalignment. Actually this is necessary validation. We haven't had this before and we used to get backend error after long queuing time.

The logic is almost the copy of `ASAPScheduling` pass. This is slightly modified to avoid inserting consecutive delay, such as delay(100dt)-delay(12dt)-measure. To isolate the context of `alignment` from other passes, the `ASAPSchedule` pass is kept as-is.

ajavadia commented on qiskit/compiler/transpiler.py:
```diff
@@ -147,6 +148,11 @@ def transpile(
+            The location of specific instructions is adjusted to be at quantized time that is
+            multiple of this value. This information will be provided by the backend configuration.
+            If the backend doesn't have any restriction on the alignment,
+            then this adjustment routine will be ignored.
```
```suggestion
            then ``alignment`` is None and no adjustment will be performed.
```

ajavadia commented on test/python/transpiler/test_instruction_alignments.py:
```diff
@@ -0,0 +1,314 @@
+        c: 1/════════════════════════╩═
+                                     0
+
+        This type of experiment slightly change delay duration of interest.
```
```suggestion
        This type of experiment slightly changes delay duration of interest.
```

ajavadia commented on test/python/transpiler/test_instruction_alignments.py:
```diff
@@ -0,0 +1,314 @@
+        «c: 2/═══════════════════
+        «
+
+        Delay for the another channel paired by multi-qubit instruction is also scheduled.
```
```suggestion
        Delay for the other channel paired by multi-qubit instruction is also scheduled.
```

ajavadia commented on test/python/transpiler/test_instruction_alignments.py:
```diff
@@ -0,0 +1,314 @@
+        «
+
+        Delay for the another channel paired by multi-qubit instruction is also scheduled.
+        Delay (1782dt) = X (160dt) + Delay (100dt + extra 12dt) + Measure (1600dt).
```
```suggestion
        Delay (1872dt) = X (160dt) + Delay (100dt + extra 12dt) + Measure (1600dt).
```

ajavadia commented on releasenotes/notes/add-alignment-management-passes-650b8172e1426a73.yaml:
```diff
@@ -0,0 +1,15 @@
+    set arbitrary alignment value directly to the transpiler.
+    If backend doesn't provide alignment information, then this value defaults to 1 and
+    execution of these passes are just skipped.
+    Backends assigned to IBM Quantum Servies will shortly provide this new information to
```
```suggestion
    Backends assigned to IBM Quantum Services will shortly provide this new information to
```

ajavadia commented:
> The logic is almost the copy of ASAPScheduling pass.

Instead of copying ASAPScheduling, isn't it easier to just iterate over all `Measure` nodes, take their earliest predecessor `Delay`, and increase it by some amount? If not `Delay` predecessor node is found, create one.

ajavadia commented on test/python/transpiler/test_instruction_alignments.py:
```diff
@@ -0,0 +1,314 @@
+
+        (input)
+
+             ┌────┐┌────────────────┐┌───┐┌────────────────┐┌────┐┌─┐
+        q_0: ┤ √X ├┤ Delay(100[dt]) ├┤ X ├┤ Delay(100[dt]) ├┤ √X ├┤M├
+             └────┘└────────────────┘└───┘└────────────────┘└────┘└╥┘
+        c: 1/══════════════════════════════════════════════════════╩═
+                                                                   0
+
+        (output)
+
+             ┌────┐┌────────────────┐┌───┐┌────────────────┐┌────┐┌──────────────┐┌─┐
+        q_0: ┤ √X ├┤ Delay(100[dt]) ├┤ X ├┤ Delay(100[dt]) ├┤ √X ├┤ Delay(8[dt]) ├┤M├
+             └────┘└────────────────┘└───┘└────────────────┘└────┘└──────────────┘└╥┘
+        c: 1/══════════════════════════════════════════════════════════════════════╩═
+                                                                                   0
```
this is interesting but I don't really understand how the pass decides to insert a delay here, rather than just increase the second `Delay(100[dt])`

nkanazawa1989 replied:
If the pass find misalignment of measurement, this just insert the delay in between previous instruction and itself. Meaning SX and measure here. Note that delay is also an instruction, therefore if we have Delay(x dt)-Measure then this will become Delay(x + dx dt)-Measure.

nkanazawa1989 commented:
>  just iterate over all Measure nodes

Yes this is what I wanted to implement, however we cannot do this because DAGNode doesn't have information of t0 even after scheduled. Perhaps we need to allow a node to have t0 after this release.

itoko commented:
In addition, if we have a mid measurement, the insertion of delay in front of the measurement affects the start times of the following gates globally, suggesting local changes on measurements would not work for circuits with mid measurements.

ajavadia commented on qiskit/compiler/transpiler.py:
```diff
@@ -147,6 +148,21 @@ def transpile(
+            .. note::
+
+                Note that currently this instruction time optimization is performed only for
+                ``measure`` gate instructions. Here we assume a hardware that
+                ``measure`` gate should start from time which is multiple of the alignment value.
```
```suggestion
                ``measure`` instructions. Here we assume a hardware whose
                measurements should start at a time which is a multiple of the alignment value.
```

ajavadia reviewed (commented):
ok this looks good to me now. thanks for the explanations and making the updates.
